modules/libcom/RTEMS/posix/rtems_init.c

```diff
@@ -0,0 +1,1174 @@
+                                           , "set timezone (obsolete?)"
+#endif
+                                           };
+static void zonesetCallFunc(const iocshArgBuf *args)
+{
+    zoneset(args[0].sval);
+}
+
+
+/*
+ * Register RTEMS-specific commands
+ */
+static void iocshRegisterRTEMS (void)
+{
+    iocshRegister(&netStatFuncDef, netStatCallFunc);
+    iocshRegister(&heapSpaceFuncDef, heapSpaceCallFunc);
+#ifndef OMIT_NFS_SUPPORT
+    iocshRegister(&nfsMountFuncDef, nfsMountCallFunc);
+#endif
+    iocshRegister(&zonesetFuncDef, &zonesetCallFunc);
+    iocshRegister(&rtshellFuncDef, &rtshellCallFunc);
+#if __RTEMS_MAJOR__ > 4
+    rtems_shell_init_environment();
+#endif
+}
+
+/*
+ * Set up the console serial line (no handshaking)
+ */
+static void
+initConsole (void)
+{
+    struct termios t;
+
+printf("\n initConsole --- Info ---\n");
+printf("stdin: fileno: %d, ttyname: %s\n", fileno(stdin), ttyname(fileno(stdin)));
+printf("stdout: fileno: %d, ttyname: %s\n", fileno(stdout), ttyname(fileno(stdout)));
+printf("stderr: fileno: %d, ttyname: %s\n", fileno(stderr), ttyname(fileno(stderr)));
+
+    if (tcgetattr (fileno (stdin), &t) < 0) {
+        printf ("tcgetattr failed: %s\n", strerror (errno));
+        return;
+    }
+    t.c_iflag &= ~(IXOFF | IXON | IXANY);
+    if (tcsetattr (fileno (stdin), TCSANOW, &t) < 0) {
+        printf ("tcsetattr failed: %s\n", strerror (errno));
+        return;
+    }
+}
+
+/*
+ * Hook to ensure that BSP cleanup code gets run on exit
+ */
+static void
+exitHandler(void)
+{
+    rtems_shutdown_executive(0);
+}
+
+#ifndef RTEMS_LEGACY_STACK
+static char* getPrimaryNetworkInterface(void)
+{
+    // lookup available network interfaces
+    char ifnamebuf[IF_NAMESIZE];
+    char *ifname;
+    // get primary network interface
+    ifname = if_indextoname(1, &ifnamebuf[0]);
+    if (ifname == NULL) return (NULL);
+    printf("\n***** Primary Network interface : %s *****\n", ifname);
+    return (ifname);
+}
+
+static void
+dhcpcd_hook_handler(rtems_dhcpcd_hook *hook, char *const *env)
+{
+    int bound = 0;
+    char iName[16];
+    char *name;
+    char *value;
+    char * env_position;
+    
+	(void)hook;
+
+    char ifnamebuf[IF_NAMESIZE];
+    sprintf(ifnamebuf, "%s", getPrimaryNetworkInterface());
+
+    while (*env != NULL) {
+        char const * interface = "interface";
+        char const * reason = "reason";
+        char const * bound_str = "BOUND";
+        
+        name = strtok_r(*env,"=", &env_position);
+        value = strtok_r(NULL,"=", &env_position);
+        printf("all out ---> %s = %s\n", name, value);
+
+        if (!strncmp(name, interface, strlen(interface)) &&
+            !strcmp(value,  ifnamebuf)) {
+            snprintf(iName, sizeof(iName), "%s", value);
+        }
+
+        if (!strncmp(name, reason, strlen(reason)) &&
+            !strncmp(value, bound_str, strlen(bound_str))) {
+            printf ("Interface %s bounded\n", iName);
+            bound = 1;
+        }
```

Comment:
@hjunkes, This code is searching for "reason=BOUND" in the dhcpcd hook environment, but I haven't seen that key/value pair in my runs.  If you search the following example output for "all out", there are no "all out reason=BOUND" lines:  https://github.com/epics-base/epics-base/files/6005278/mvme2500-libcom.txt

Was there some documentation that indicated it should send result=BOUND sometimes?

Rather than search for "result=BOUND", do you think we should search for at least one interface with "if_up=true" instead, to know that one of the interfaces is up?

Reply:
fyi. `IF_UP` does not imply completion.  While a DHCP client is searching an interface will by "up", but have no address assigned, and thus not be usable.

One place to start would be printing the complete `env` before the loop.

Reply:
Exactly, that is the reason. This also works fine for me. The messages can also go away at some point. I'm also already in the process of what the presets of the argv's to change.
Unfortunately I failed grandiosely when I tried everything with RTEMS6 :-(

But a nice Makefile for creating RTEMS here:
https://gitlab.com/c-mauderer/rtems-simulation-environment

Reply:
Unless I mistake your meaning, the complete `env` is already being printed out as the messages starting with "all out".  The output below shows my qemu session and three sections of environment printed out for each of the three times rtems_dhcpcd_hook callback is called.

***** Primary Network interface : re0 *****
all out ---> interface = re0
all out ---> reason = PREINIT
all out ---> pid = 1
all out ---> ifmetric = 201
all out ---> ifwireless = 0
all out ---> ifflags = 4294936643
all out ---> ifmtu = 1500
all out ---> interface_order = re0
all out ---> if_up = false
all out ---> if_down = true

***** Primary Network interface : re0 *****
all out ---> interface = re0
all out ---> reason = CARRIER
all out ---> pid = 1
all out ---> ifmetric = 201
all out ---> ifwireless = 0
all out ---> ifflags = 4294936643
all out ---> ifmtu = 1500
all out ---> interface_order = re0
all out ---> if_up = false
all out ---> if_down = true


***** Primary Network interface : re0 *****
all out ---> interface = re0
all out ---> reason = IPV4LL
all out ---> pid = 1
all out ---> ifmetric = 201
all out ---> ifwireless = 0
all out ---> ifflags = 4294936643
all out ---> ifmtu = 1500
all out ---> interface_order = re0
all out ---> if_up = true
all out ---> if_down = false
all out ---> new_ip_address = 169.254.163.246
all out ---> new_subnet_cidr = 16
all out ---> new_network_number = 169.254.0.0
all out ---> new_subnet_mask = 255.255.0.0
all out ---> new_broadcast_address = 169.254.255.255

Reply:
Ah, I wasn't reading closely enough this morning.

Looking at `make_env()` in `dhcpcd/script.c` from libbsd it looks like entries in this array may be relied upon to have a nil terminator.  So one way to simplify (or at least condense) this loop might be:

```c
for(; *env; env++) {
    const char *val;
#define IF(NAME) if(strncmp(*env, NAME "=", 1+strlen(NAME))==0 && (val=1+strlen(NAME)+*env))

    IF("interface") {
    }
    IF("reason") {
        printf("reason is '%s'\n", val);
    }
    ...
#undef IF
}
```

Reply:
```
all out ---> new_ip_address = 169.254.163.246
all out ---> new_subnet_cidr = 16
all out ---> new_network_number = 169.254.0.0
all out ---> new_subnet_mask = 255.255.0.0
all out ---> new_broadcast_address = 169.254.255.255
```
@brendanchandler That IP address is in the zeroconf range so it probably didn't come from dhcp at all. We will need to turn off DHCP and configure the network of our MVME2500 boards directly (which is why I'd like RTEMS to read its settings from U-Boot), but I have other stuff I have to work on now – let me know if you want access to the board.

Reply:
> Was there some documentation that indicated it should send reason=BOUND sometimes?

fyi. The `env` array is exporting the same key/value pairs normally exported to dhcpcd hook scripts on Linux as environment variables.  cf. `man dhcpcd/dhcpcd-run-hooks.8.in` for a list of hooks.  Also `man dhcpcd/dhcpcd.conf.5.in`.